# =============================================================================
# plot_example.R
#
# Example script demonstrating how the output files of the connectivity
# enrichment analysis can be loaded and visualized.
# (1) Load AUCs and compute p-values / scores (cf. Figs. 5,6 of the paper)
# (2) Plot enrichment curves (cf. Fig. 5a of the paper)
#
# The script is provided "as is" under the MIT license included with the magnum
# distribution.
#
# --
# Daniel Marbach
# April 27, 2015
# =============================================================================


# -----------------------------------------------------------------------------
# Initialization

# Clean up
#rm(list=ls())
#dev.off()

# Load functions
source("plot_functions.R")

# Here we just load results from one GWAS and network kernel, but the provided
# functions also work with vectors specifying several GWASs and/or kernels.
# The following result files are loaded, which can be generated by following
# the step-by-step tutorial in the magnum user guide:
# - macular_degeneration_neovascular--smooth_muscle_cells_-_umbilical_vein_4stepKernel_alpha2.0_weighted.AUC.txt.gz
# - macular_degeneration_neovascular--smooth_muscle_cells_-_umbilical_vein_4stepKernel_alpha2.0_weighted.txt.gz

# The name of the GWASs
gwas <- "macular_degeneration_neovascular"
# The name of the kernels
kernels <- "smooth_muscle_cells_-_umbilical_vein_4stepKernel_alpha2.0_weighted"
# The directory where result files are located (filenames are expected to be: 
# <gwas>--<kernel>.txt.gz, which is how magnum names the result files). Here we assume
# that the result files are located in the parent directory of the R working directory.
dir <- ".."


# -----------------------------------------------------------------------------
# (1) Load AUCs and compute p-values / scores

# Load AUCs
aucs <- loadAllAUCs(dir=dir, gwas=gwas, kernels=kernels, numPermut=10000, k=2, logScale=TRUE)
# Compute empirical p-values
pvals <- computeEnrichmentPvals(aucs)
# Benjamini-Hochberg correction
#pvals <- fdrCorrection(pvals)
# Compute scores
scores <- -log10(pvals)

# Print results
cat("p-value = ", pvals, "\n")
cat("score   = ", scores, "\n\n")


# -----------------------------------------------------------------------------
# (2) Plot connectivity enrichment curves

# Fraction of the enrichment curve included in the data files (value of the magnum
# option "--curve"). Used to correctly label x-axis and compute number of genes
# passing the genome-wide signifcance threshold in the GWAS. 
curveCutoff <- 0.2

# x-range for the enrichment curve plot (set to NULL to plot the complete curve)
xrange <- c(0,1000)
# y-range for the enrichment curve plot
yrange <- c(-0.2,0.2)
# Name of the output file
outfile <- "enrichmentCurve.pdf"

# Plot curves and compute significant number of enriched genes
sg <- generateEnrichmentCurvePlots(dir=dir, gwas=gwas, kernels=kernels,
                                   xrange=xrange, yrange=yrange,
                                   curveCutoff=curveCutoff,
                                   pvals=pvals,
                                   outfile=outfile)

